### PR TITLE
Extend excel class to do population

### LINKED ISSUE #6: Extends service class to handle excel population for one IO [1~2 day]
Extends the existing query submission service class so that it can use the returned results to populate the spreadsheet accordingly.

### PR DESCRIPTION
**Pull request by hanj7221**
Closing #6

New issues that arose in this PR and will be resolved in a separate PR:

- Conditional format for the budget amount column: #34 
- Only sum up active budget segments: #35 
- Name for new sheet needs to be unique: #36

### CODE REVIEW

File: lib/src/excel.dart [hunk 1]
```diff
@@ -1,36 +1,128 @@
...
 class ExcelDart {
-  ExcelDart._private();
+  static final _startAddressInt = 'A'.codeUnitAt(0);
+  static final _fontName = 'Roboto';
+  static final _fontSize = 12;
+  static final _horizontalAlignment = 'Center';
+  static final _borderStyle = 'Continuous';
 
-  static final ExcelDart _singleton = ExcelDart._private();
+  static final _tableHeader = [
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:25]: does this have to keep in line with the row proto? what happens if they mismatch?

File: lib/src/excel.dart [hunk 2]
```diff
@@ -1,36 +1,128 @@
...
 
-  factory ExcelDart() {
-    return _singleton;
-  }
-
-  /// Waits for Office APIs to be ready and then executes function [_populate].
-  void exec() async {
+  /// Waits for Office APIs to be ready and then populates the spreadsheet with
+  /// entries in the [insertionOrderList]
+  static void populate(List insertionOrderList) async {
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: add type parameter

File: lib/src/excel.dart [hunk 3]
```diff
@@ -1,36 +1,128 @@
...
     await Office.onReady(allowInterop((info) async {
-      await ExcelJS.run(allowInterop(_populate));
+      await ExcelJS.run(allowInterop((context) {
+        // Grabs the current worksheet
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: nit: comment should be formatted as sentences

File: lib/src/excel.dart [hunk 4]
```diff
@@ -1,36 +1,128 @@
...
+        final sheet = context.workbook.worksheets.getActiveWorksheet();
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: consider: making a new sheet to avoid problems of overwriting/conflicting with existing data

File: lib/src/excel.dart [hunk 5]
```diff
@@ -1,36 +1,128 @@
...
+          ..format.borders.getItem('EdgeBottom').style = _borderStyle;
+
+        // Adds and formats table body
+        table.getDataBodyRange()
+          ..formulas = _generateTableBody(insertionOrderList)
+          ..format.font.name = _fontName
+          ..format.font.size = _fontSize
+          ..format.horizontalAlignment = _horizontalAlignment;
+
+        // Auto-fits all used cells
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: love the comments especially as a reviewer not familiar with excel api

File: lib/src/excel.dart [hunk 6]
```diff
@@ -1,36 +1,128 @@
...
-    final values = <List<String>>[
-      ['test test']
-    ];
-    range.values = values;
-    range.format.autofitColumns();
-    return context.sync();
+  static List _generateTableBody(List insertionOrderList) {
+    final tableBody = [];
+
+    for (InsertionOrder io in insertionOrderList) {
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: what about doing something like insertionOrderList.map(conversionfunction).toList() do you think that would be more idiomatic?

File: lib/src/excel.dart [hunk 7]
```diff
@@ -1,36 +1,128 @@
...
+        _getEndDate(io.budget.budgetSegments.last),
+      ]);
+    }
+    return tableBody;
+  }
+
+  static String _calculateTotalBudgetAmount(
+      List<InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment>
+          budgetSegments) {
+    var totalBudgetMicros = Int64.ZERO;
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: nice use of the constant

File: lib/src/excel.dart [hunk 8]
```diff
@@ -1,36 +1,128 @@
...
+
+    for (var segment in budgetSegments) {
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: what about using a reducer?

File: lib/src/excel.dart [hunk 9]
```diff
@@ -1,36 +1,128 @@
...
+      totalBudgetMicros += Int64.parseInt(segment.budgetAmountMicros);
+    }
+
+    return (totalBudgetMicros.toDouble() / 1000000).toString();
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: is the toDouble necessary for division? that could result in some precision loss
consider storing the 1000000 as a constant. or having a conversion function
also does this format it like we would expect money to be formatted? ie $##.## ?

**Review by ruben-aghayan — CHANGES_REQUESTED** [review]
since we dont have a testing story for the pieces interacting with excel, consider writing (brief) instructions for manual testing with some expectations

File: lib/src/excel.dart (hunk 9 above)
Comment (arulprakash-g) [reply lib/src/excel.dart]: Also currency vs impression budgets?

File: lib/src/excel.dart [hunk 10]
```diff
@@ -1,36 +1,128 @@
...
+        io.budget.automationType.toString(),
+        _calculateTotalBudgetAmount(io.budget.budgetSegments),
+        _getStartDate(io.budget.budgetSegments.first),
+        _getEndDate(io.budget.budgetSegments.last),
+      ]);
+    }
+    return tableBody;
+  }
+
+  static String _calculateTotalBudgetAmount(
```
Comment (arulprakash-g) [inline lib/src/excel.dart:115]: I think this is not very relevant. Advertisers are mostly interested in their currently active segment.

File: lib/src/excel.dart (hunk 9 above)
Comment (arulprakash-g) [reply lib/src/excel.dart]: Also impression vs currency budget.

File: lib/src/excel.dart [hunk 11]
```diff
@@ -1,36 +1,128 @@
...
+      tableBody.add([
+        io.insertionOrderId,
+        io.advertiserId,
+        io.campaignId,
+        io.displayName,
+        io.entityStatus.toString(),
+        io.updateTime,
+        io.pacing.pacingPeriod.toString(),
+        io.pacing.pacingType.toString(),
+        io.pacing.dailyMaxMicros,
```
Comment (arulprakash-g) [inline lib/src/excel.dart]: This is in micros and total budget we divide by 10^6 - lets use a consistent unit.

**Review by arulprakash-g — APPROVED** [review]

File: lib/src/excel.dart (hunk 9 above)
Comment (hanj7221) [reply lib/src/excel.dart]: The formatting will be done from the Excel side using their javascript APIs. Conditional format the budget amount column based on the budgetType (currency or impression) will be done in a separate PR. New issue created #34.

File: lib/src/excel.dart (hunk 1 above)
Comment (hanj7221) [reply lib/src/excel.dart:25]: tableHeader will need to be in line with the _generateTableRow() function. Changing the protobuff definitions can affect the _generateTableRow() function (invalid getters etc). I will add a comment specifying this.

File: lib/src/excel.dart (hunk 10 above)
Comment (hanj7221) [reply lib/src/excel.dart:115]: New issue created #35, will be addressed in a separate PR.

File: lib/src/excel.dart [hunk 12]
```diff
@@ -33,20 +40,26 @@ class ExcelDart {
...
+  /// a new spreadsheet with [sheetName] and populates the spreadsheet with
   /// entries in the [insertionOrderList]
-  static void populate(List insertionOrderList) async {
+  static void populate(List<InsertionOrder> insertionOrderList) async {
     await Office.onReady(allowInterop((info) async {
       await ExcelJS.run(allowInterop((context) {
-        // Grabs the current worksheet
-        final sheet = context.workbook.worksheets.getActiveWorksheet();
+        // Adds a new worksheet
+        final sheet = context.workbook.worksheets.add('Query');
```
Comment (hanj7221) [inline lib/src/excel.dart:50]: Each new sheet needs a unique name. New issue created for this #36, and will be addressed in a separate PR.

File: lib/src/excel.dart (hunk 9 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart]: i think we want to handle currency budget case end to end first then impression budget?

File: lib/src/excel.dart (hunk 10 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart:115]: ok lets make sure #35 doesnt fall off since its part of our mvp

File: lib/src/excel.dart [hunk 13]
```diff
@@ -1,36 +1,139 @@
...
+  static final _startAddressInt = 'A'.codeUnitAt(0);
+  static final _fontName = 'Roboto';
+  static final _fontSize = 12;
+  static final _horizontalAlignment = 'Center';
+  static final _borderStyle = 'Continuous';
 
-  static final ExcelDart _singleton = ExcelDart._private();
+  /// Header for the master table.
+  ///
+  /// If changes have been to [insertion_order.proto] or
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: have been made to

File: lib/src/excel.dart [hunk 14]
```diff
@@ -1,36 +1,139 @@
...
-  Future<void> _populate(RequestContext context) async {
-    final sheet = context.workbook.worksheets.getActiveWorksheet();
-    final range = sheet.getRange('A1');
-    final values = <List<String>>[
-      ['test test']
-    ];
-    range.values = values;
-    range.format.autofitColumns();
-    return context.sync();
+  static List _generateTableRow(InsertionOrder io) => [
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: add type parameter

File: lib/src/excel.dart [hunk 15]
```diff
@@ -1,36 +1,139 @@
...
+
+  static String _calculatePacingDailyMax(String dailyMaxMicros) =>
+      dailyMaxMicros.isEmpty
+          ? dailyMaxMicros
+          : (Int64.parseInt(dailyMaxMicros) * 1e-6).toString();
+
+  static String _calculateTotalBudgetAmount(
+      List<InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment>
+          budgetSegments) {
+    var budgetAmountList = budgetSegments
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: can this be final?

File: lib/src/excel.dart [hunk 16]
```diff
@@ -1,36 +1,139 @@
...
+        .map((segment) => Int64.parseInt(segment.budgetAmountMicros))
+        .toList();
+    var totalBudgetMicros =
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: same

File: lib/src/excel.dart (hunk 10 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart:115]: you can drop a todo with the issue on the method

File: lib/src/excel.dart [hunk 17]
```diff
@@ -1,36 +1,139 @@
...
+        budgetAmountList.reduce((value, element) => value + element);
+
+    return (totalBudgetMicros.toDouble() * 1e-6).toString();
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:124]: same with leaving a todo to handle the formatting

**Review by ruben-aghayan — APPROVED** [review]

File: lib/src/excel.dart (hunk 16 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart]: can you merge these list comps?
reduce (value + Int64(parseInt(element.budgetAmountMicros))

Comment (hanj7221) [reply lib/src/excel.dart]: Reducer needs to return the same type as the list, but now the map & reduce are chained together so we don't the intermediate variable.
